perllib/FixMyStreet/App/Controller/Admin/Waste.pm

```diff
@@ -0,0 +1,112 @@
+package FixMyStreet::App::Controller::Admin::Waste;
+
+use JSON::MaybeXS;
+use Moose;
+use Try::Tiny;
+use namespace::autoclean;
+
+BEGIN { extends 'Catalyst::Controller'; }
+
+=head1 NAME
+
+FixMyStreet::App::Controller::Admin::Waste - Catalyst Controller
+
+=head1 DESCRIPTION
+
+Admin pages for configuring WasteWorks parameters
+
+=head1 METHODS
+
+=cut
+
+sub index : Path : Args(0) {
+    my ( $self, $c ) = @_;
+
+    my $user = $c->user;
+
+    if ($user->is_superuser) {
+        $c->forward('fetch_wasteworks_bodies');
+    } elsif ( $user->from_body ) {
+        $c->forward('load_wasteworks_body', [ $user->from_body->id ]);
+        $c->res->redirect( $c->uri_for_action( 'admin/waste/edit', $c->stash->{body}->id ) );
+    } else {
+        $c->detach( '/page_error_404_not_found', [] );
+    }
+}
+
+sub edit : Path : Args(1) {
+    my ( $self, $c, $body_id ) = @_;
+
+    foreach (qw(status_message)) {
+        $c->stash->{$_} = $c->flash->{$_} if $c->flash->{$_};
+    }
+
+    $c->forward('load_wasteworks_body', [ $body_id ]);
+    $c->forward('stash_body_config_json');
+    $c->forward('/auth/get_csrf_token');
+
+    if ($c->req->method eq 'POST') {
+        $c->forward('/auth/check_csrf_token');
```

Comment:
Hasn't this already happened above, or do you always need to do it again for POSTs?

Reply:
Above we call `get_csrf_token` which sets `csrf_time` and `csrf_token` in the stash. This call to `check_csrf_token` is only needed on POSTs because it compares the value that was sent by the user with the expected value and errors if they don't match.